perfkitbenchmarker/vpn_service.py

```diff
@@ -0,0 +1,326 @@
+                     'Number of tunnels to create for each VPN GW pair.')
+flags.DEFINE_integer('vpn_service_gateway_count', None,
+                     'Number of VPN GWs to create for each vm_group.')
+flags.DEFINE_string('vpn_service_name', None,
+                    'If set, use this name for VPN Service.')
+flags.DEFINE_string('vpn_service_shared_key', None,
+                    'If set, use this PSK for VPNs.')
+flags.DEFINE_string('vpn_service_routing_type', None,
+                    'static or dynamic(BGP)')
+flags.DEFINE_integer('vpn_service_ike_version', None, 'IKE version')
+
+FLAGS = flags.FLAGS
+
+
+def GetVPNServiceClass():
+  """Gets the VPNService class.
+
+  Args:
+
+  Returns:
+    Implementation class
+  """
+  return resource.GetResourceClass(VPNService)
+
+
+class VPN(object):
+  """An object representing the VPN.
+
+  A VPN instance manages tunnel configurations for exactly 1 pair of endpoints.
+  """
+
+  def __init__(self, *args, **kwargs):
+      return object.__init__(self, *args, **kwargs)
+
+  def getKeyFromGWPair(self, gateway_pair, suffix=''):
+    """Return the VPN key for a pair of endpoints.
+
+    Args:
+      gateway_pair: A tuple of 2 VPN gateways which define the VPN tunnel.
+      suffix: A unique suffix if multiple tunnels b/t this gateway pair exist.
+
+    Returns:
+      string. The VPN key.
+
+    """
+    key = 'vpn' + ''.join(gw for gw in gateway_pair) + suffix + FLAGS.run_uri
+    return key
+
+  def Create(self, gateway_pair, suffix=''):
+    self.GWPair = gateway_pair
+    self.name = self.getKeyFromGWPair(gateway_pair)
+    self.tunnel_config = TunnelConfig(tunnel_name=self.name, suffix=suffix)
+
+  def Delete(self):
+    pass
+
+  def GetVPN(self, gateway_pair, suffix=''):
+    """Gets a VPN object for the gateway_pair or creates one if none exists
+
+    Args:
+    gateway_pair: a tuple of two VPNGWs
+    """
+
+    benchmark_spec = context.GetThreadBenchmarkSpec()
+    if benchmark_spec is None:
+      raise errors.Error('GetVPN called in a thread without a '
+                         'BenchmarkSpec.')
+    with benchmark_spec.vpns_lock:
+      key = self.getKeyFromGWPair(gateway_pair, suffix)
+      if key not in benchmark_spec.vpns:
+        self.Create(gateway_pair, suffix)
+        benchmark_spec.vpns[key] = self
+      return benchmark_spec.vpns[key]
+
+  def ConfigureTunnel(self):
+    """Configure the VPN tunnel."""
+
+    benchmark_spec = context.GetThreadBenchmarkSpec()
+    vpn_gateway_0 = benchmark_spec.vpn_gateways[self.GWPair[0]]
+    vpn_gateway_1 = benchmark_spec.vpn_gateways[self.GWPair[1]]
+
+    assert not (vpn_gateway_0.require_target_to_init and vpn_gateway_1.require_target_to_init), 'Cant connect 2 passive VPN GWs'
+
+    tunnel_config_hash = None
+    while not self.isTunnelConfigured():
+        vpn_gateway_0.ConfigureTunnel(self.tunnel_config)
+        vpn_gateway_1.ConfigureTunnel(self.tunnel_config)
+        if self.tunnel_config.hash() == tunnel_config_hash:
```

Comment:
Add a comment here as I'm not sure what's going on.  It looks like the self.tunnel_config.hash()  should change with every call?  I'm not sure why this won't result in a ValueError the 2nd time through.

Reply:
Added a comment above to describe what's happening. The hash() was added after the [previous comment](https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/pull/2083#discussion_r398106992) when I realized the config loop didn't actually exit when no more information was added. The tunnel_config should be updated by at least one endpoint on each pass. If both endpoints have enough data to setup their tunnel then the loop ends normally and the tunnels are created. If not enough information is available and neither endpoint can add anything then we raise the ValueError here.